src/internal/kopia/upload.go

```diff
@@ -754,30 +754,49 @@ func inflateCollectionTree(
 			}
 
 			updatedPaths[s.PreviousPath().String()] = s.FullPath()
-		}
 
-		// TODO(ashmrtn): Get old location ref and add it to the prefix matcher.
-		lp, ok := s.(data.LocationPather)
-		if ok && s.PreviousPath() != nil {
-			if err := toMerge.addLocation(s.PreviousPath().ToBuilder(), lp.LocationPath()); err != nil {
-				return nil, nil, clues.Wrap(err, "building updated location set").
-					With("collection_location", lp.LocationPath())
+			// Safe to do only for moved collections as we really only need prefix
+			// matching if a nested folder changed in some way that didn't generate a
+			// collection. For that to the be case, the nested folder's path must have
+			// changed via one of the ancestor folders being moved. This catches the
+			// ancestor folder move.
+			lp, ok := s.(data.PreviousLocationPather)
+			if ok {
+				prevLoc := lp.PreviousLocationPath()
+				newLoc := lp.LocationPath()
+
+				if prevLoc == nil {
+					return nil, nil, clues.New("moved collection with nil previous location").
+						With("collection_full_path", s.FullPath())
```

Comment:
minor: `collection_full_path` (and `collection_previous_path` elsewhere) seem to be duplicated a couple times.  Might be worth moving into an inner ctx within the loop?